clusterloader2/pkg/prometheus/gce_windows_util.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prometheus
+
+import (
+	"fmt"
+	"os/exec"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/pkg/config"
+	"k8s.io/perf-tests/clusterloader2/pkg/framework/client"
+)
+
+// gceNode is a struct storing gce node info
+type gceNode struct {
+	projectID string
+	zone      string
+	nodeName  string
+}
+
+func setUpWindowsNodeAndTemplate(k8sClient kubernetes.Interface, mapping map[string]interface{}) error {
+	// Get the gce windows node info
+	windowsNode, err := getGceWindowsNodeFromKubernetesService(k8sClient)
+	if err != nil {
+		return err
+	}
+	// Install the wmi exporter onto windows node
+	if err := installWmiExporter(k8sClient, windowsNode, mapping); err != nil {
+		return err
+	}
+	mapping["WINDOWS_NODE_NAME"] = windowsNode.nodeName
+	mapping["PROMETHEUS_SCRAPE_WINDOWS_NODES"] = true
+	return nil
+}
+
+func enableScrapingWindowsNode(mapping map[string]interface{}, clusterLoaderConfig *config.ClusterLoaderConfig) bool {
+	if windowsNodeTest, exists := mapping["WINDOWS_NODE_TEST"]; exists && windowsNodeTest.(bool) && clusterLoaderConfig.ClusterConfig.Provider == "gce" {
+		return true
+	}
+	return false
+}
+
+func installWmiExporter(k8sClient kubernetes.Interface, windowsNode *gceNode, mapping map[string]interface{}) error {
+	wmiExporterURL, ok := mapping["WMI_EXPORTER_URL"]
+	if !ok {
+		return fmt.Errorf("missing setting up wmi exporter download url")
+	}
+	wmiExporterCollectors, ok := mapping["WMI_EXPORTER_ENABLED_COLLECTORS"]
+	if !ok {
+		return fmt.Errorf("missing setting up wmi exporter enabled collectors")
+	}
+	installWmiCmdTemplate := `Start-Job -Name InstallWmiExporter -ScriptBlock {invoke-webrequest -uri %s -outfile C:\wmi_exporter.msi; Start-Process msiexec.exe -Wait -ArgumentList '/I C:\wmi_exporter.msi ENABLED_COLLECTORS=%s LISTEN_PORT=5000 /quiet'}; Wait-Job -Timeout 300 -Name InstallWmiExporter`
```

Comment:
You could add a comment here explaining what you found: that the only way you could get this to work was using -ScriptBlock, executing it directly was failing for some reason, etc.

Reply:
Right, added